pnpm_{8,9,10}: add updateScript; pnpm_10: 10.6.3 -> 10.6.5

Scrumplex opened the pull request:
https://github.com/pnpm/pnpm/releases/tag/v10.6.5

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Scrumplex committed:
pnpm_{8,9,10}: add updateScript

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex commented:
Sadly this will cause a rebuild of all packages using pnpm, as `pname` has changed

SuperSandro2000 commented on pkgs/development/tools/pnpm/generic.nix:
```diff
@@ -75,6 +79,28 @@ stdenvNoCC.mkDerivation (finalAttrs: {
       tests.version = lib.optionalAttrs withNode (
         testers.testVersion { package = finalAttrs.finalPackage; }
       );
+      updateScript = writeScript "discord-update-script" ''
```
```suggestion
      updateScript = writeScript "pnpm-update-script" ''
```

SuperSandro2000 commented on pkgs/development/tools/pnpm/generic.nix:
```diff
@@ -75,6 +79,28 @@ stdenvNoCC.mkDerivation (finalAttrs: {
+        #!nix-shell -i bash -p curl jq common-updater-scripts
+        set -eou pipefail;
+
+        curl_github() {
+            curl -L ''${GITHUB_TOKEN:+" -u \":$GITHUB_TOKEN\""} "$@"
+        }
+
+        majorVersion=""
+
+        latestTag=$(curl_github https://api.github.com/repos/pnpm/pnpm/releases?per_page=100 | jq -r --arg major "${majorVersion}" '[.[].tag_name | select(startswith("v" + $major))][0]')
```
```suggestion
        latestTag=$(curl -L ''${GITHUB_TOKEN:+" -u \":$GITHUB_TOKEN\""} https://api.github.com/repos/pnpm/pnpm/releases?per_page=100 | jq -r --arg major "${majorVersion}" '[.[].tag_name | select(startswith("v" + $major))][0]')
```

gepbird replied:
I think this makes it harder to read

SuperSandro2000 replied:
First of all majorVersion is unused. We also don't need to hide things in functions for no reason.

gepbird reviewed (changes requested):
Thanks for adding an update script!

gepbird commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 390687`

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 9 packages built:</summary>
  <ul>
    <li>aonsoku</li>
    <li>authelia</li>
    <li>fedistar</li>
    <li>pnpm (pnpm_10)</li>
    <li>renovate</li>
    <li>vencord</li>
    <li>vencord-web-extension</li>
    <li>webcord-vencord</li>
    <li>youtube-music</li>
  </ul>
</details>


(I started this before you added the update script, hence the lower rebuild number)

gepbird reviewed (approved):
Tested the update script for both versions, built a pnpm 8, 9 and 10 package, and previously built all pnpm 10 packages. Changes LGTM

SuperSandro2000 commented on pkgs/development/tools/pnpm/generic.nix:
```diff
@@ -75,6 +79,27 @@ stdenvNoCC.mkDerivation (finalAttrs: {
+
+        latestVersion="''${latestTag#v}"
+
+        update-source-version ${finalAttrs.pname} "$latestVersion" --file=./pkgs/development/tools/pnpm/default.nix
```
```suggestion
        update-source-version pnpm_${majorVersion} "$latestVersion"
```
pname only matches now by accident. We really want to match the attr of the package which has no relation to pname. Also we don't need to specify file as that would break when moving the package to by-name.

Scrumplex replied:
The path is strictly required. `meta.location` points to `generic.nix`, which does not contain the actual version number nor the hash.

SuperSandro2000 commented on pkgs/development/tools/pnpm/generic.nix:
```diff
@@ -75,6 +79,28 @@ stdenvNoCC.mkDerivation (finalAttrs: {
       );
+      updateScript = writeScript "discord-update-script" ''
+        #!/usr/bin/env nix-shell
+        #!nix-shell -i bash -p curl jq common-updater-scripts
```
I don't think setting --pure here is a good idea. It will break the script when a proxy is being used and specifying all possible envs is not possible.

Scrumplex committed:
pnpm_10: 10.6.3 -> 10.6.5

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex commented:
I have removed the pname change as suggested by @SuperSandro2000. I also reverted the pure nix-shell to accommodate proxies and other impurities.

Also updated to pnpm 10.6.5

gepbird commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 390687`

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 9 packages built:</summary>
  <ul>
    <li>aonsoku</li>
    <li>authelia</li>
    <li>fedistar</li>
    <li>pnpm (pnpm_10)</li>
    <li>renovate</li>
    <li>vencord</li>
    <li>vencord-web-extension</li>
    <li>webcord-vencord</li>
    <li>youtube-music</li>
  </ul>
</details>
